File: backend/src/main/java/com/leonardo/orgsync/orgsync/domain/entities/user/UserEntity.java
package com.leonardo.orgsync.orgsync.domain.entities.user;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.leonardo.orgsync.orgsync.domain.entities.department.Department;
import com.leonardo.orgsync.orgsync.presentation.dtos.auth.LoginRequest;
import jakarta.persistence.*;
import lombok.*;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.util.*;

@Entity
@Table(name = "tb_users")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class UserEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name= "user_id")
    private UUID id;

    @Column(nullable = false)
    private String name;
    @Column(nullable = false, unique=true)
    private String email;
    @Column(nullable = false)
    @JsonIgnore
    private String password;

    @ManyToMany(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinTable(
            name = "tb_users_role",
            joinColumns = @JoinColumn(name = "user_id"),
            inverseJoinColumns = @JoinColumn(name = "role_id")
    )
    private Set<UserRole> roles;


    @Column(nullable = false)
    private boolean enabled;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "department_id")
    @JsonBackReference
    private Department department;


    public boolean isCorretLogin(LoginRequest loginRequest, PasswordEncoder passwordEncoder) {
        return passwordEncoder.matches(loginRequest.password(), this.password);
    }
}

File: frontend/lib/domain/entities/user_entity.dart
import 'package:orgsync/domain/entities/department_entity.dart';

class UserEntity {
  final String id;
  final String name;
  final String email;
  final int roleId;
  final DepartmentEntity? department;
  bool? enabled = true;

  UserEntity({
    required this.id,
    required this.name,
    required this.email,
    required this.roleId,
    this.department,
    this.enabled,
  });

  UserEntity copyWith({
    String? id,
    String? name,
    String? email,
    DepartmentEntity? department,
    int? roleId,
    bool? enabled,
  }) {
    return UserEntity(
      id: id ?? this.id,
      name: name ?? this.name,
      email: email ?? this.email,
      department: department ?? this.department,
      roleId: roleId ?? this.roleId,
      enabled: enabled ?? this.enabled,
    );
  }
}
